torch/optim/adagrad.py

```diff
@@ -92,21 +102,157 @@ def step(self, closure=None):
             state_sums = []
             state_steps = []
 
+            has_sparse_grad = False
             for p in group['params']:
                 if p.grad is not None:
+                    if p.grad.is_sparse:
+                        has_sparse_grad = True
                     params_with_grad.append(p)
                     grads.append(p.grad)
                     state = self.state[p]
                     state_sums.append(state['sum'])
                     state_steps.append(state['step'])
 
-            F.adagrad(params_with_grad,
-                      grads,
-                      state_sums,
-                      state_steps,
-                      lr=group['lr'],
-                      weight_decay=group['weight_decay'],
-                      lr_decay=group['lr_decay'],
-                      eps=group['eps'])
+            adagrad(params_with_grad,
+                    grads,
+                    state_sums,
+                    state_steps,
+                    lr=group['lr'],
+                    weight_decay=group['weight_decay'],
+                    lr_decay=group['lr_decay'],
+                    eps=group['eps'],
+                    has_sparse_grad=has_sparse_grad,
+                    foreach=group['foreach'])
 
         return loss
+
+
+def adagrad(params: List[Tensor],
+            grads: List[Tensor],
+            state_sums: List[Tensor],
+            state_steps: List[Tensor],
+            # kwonly args with defaults are not supported by functions compiled with torchscript issue #70627
+            # setting these as kwargs for now as functional API is compiled by torch/distributed/optim
+            has_sparse_grad: bool = None,
+            foreach: bool = None,
+            *,
+            lr: float,
+            weight_decay: float,
+            lr_decay: float,
+            eps: float):
+    r"""Functional API that performs Adagrad algorithm computation.
+
+    See :class:`~torch.optim.Adagrad` for details.
+    """
+
+    if not all([isinstance(t, torch.Tensor) for t in state_steps]):
+        raise RuntimeError("API has changed, `state_steps` argument must contain a list of singleton tensors")
+
+    if foreach is None:
+        # Placeholder for more complex foreach logic to be added when value is not set
+        foreach = False
+
+    if foreach and torch.jit.is_scripting():
+        raise RuntimeError('torch.jit.script not supported with foreach optimizers')
+
+    if has_sparse_grad is None:
```

Comment:
We don't need to do that if we call into the non-multi-tensor impl. Given that this is not "free" we should skip that.
I think by moving this logic in the multi-tensor code is a good way to do that.